src/Features/Core/Portable/PullMemberUp/AbstractPullMemberUpRefactoringProvider.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PullMemberUp.QuickAction;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal abstract class AbstractPullMemberUpRefactoringProvider : CodeRefactoringProvider
+    {
+        private CancellationToken _cancellationToken;
+
+        protected abstract bool IsSelectionValid(TextSpan span, SyntaxNode userSelectedSyntax);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            // Currently support to pull field, method, event, property and indexer up,
+            // constructor, operator and finalizer are excluded.
+            _cancellationToken = context.CancellationToken;
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var userSelectedNode = root.FindNode(context.Span);
+
+
+            if (!IsSelectionValid(context.Span, userSelectedNode))
+            {
+                return;
+            }
+
+            var userSelectNodeSymbol = semanticModel.GetDeclaredSymbol(userSelectedNode);
+            if (userSelectNodeSymbol == null || userSelectNodeSymbol.ContainingType == null)
+            {
+                return;
+            }
+
+            var allTargets = FindAllValidTargets(userSelectNodeSymbol.ContainingType, context.Document.Project.Solution);
+
+            if (allTargets.Length == 0)
+            {
+                return;
+            }
+
+            if ((userSelectNodeSymbol is IMethodSymbol methodSymbol && methodSymbol.IsOrdinaryMethod()) ||
+                userSelectNodeSymbol.Kind == SymbolKind.Property ||
+                userSelectNodeSymbol.Kind == SymbolKind.Event ||
+                userSelectNodeSymbol.Kind == SymbolKind.Field)
```

Comment:
these checks should be done before you do the check to validate targets.

Reply:
should probably have a comment explaining that we're allowing the feature regardless of things like instance/static or accessibility (since you don't seem to be checking any of that).

Reply:
(or, rather, if they are checked, mention that they will be checked somewhere).